### ISSUE TITLE
Allow plugin metadata to be provided in a file instead of in the annotation

### ISSUE THREAD

**Issue by Minecrell** [SpongePowered/SpongeAPI#624]
Currently, the only standardized way to provide plugin metadata (plugin name, version, dependencies, ...) is through the `@Plugin` annotation directly on the main class. We should consider adding a way for metadata to be specified in a file additionally.

There are basically 3 ways we can consider:
- (~~**Provide no file metadata:**~~ File metadata is quite important as explained in https://github.com/SpongePowered/SpongeAPI/issues/624#issuecomment-100658625)

- **[`mcmod.info`](https://github.com/MinecraftForge/FML/wiki/FML-mod-information-file):** This is currently supported in [Sponge](https://github.com/SpongePowered/Sponge), but not in [SpongeVanilla](https://github.com/SpongePowered/SpongeVanilla) (yet?). It provides most metadata we need, although we would probably need to extend it a bit to provide all the information needed for Ore. This format is mostly standardized in the Minecraft modding platform, so we should definitely consider using this because it's also supported for a lot of mods.

- **An improved `mcmod.info` format:** This was proposed in MinecraftForge/FML#637. Especially semver or to be able to tell which version is "newer" is quite important for dependency resolution. Additionally, because this would be a new, improved format we could also standardize a few additional properties we need in Ore for an unified file format to provide mod metadata.

- (~~**A custom format only for Sponge:**~~ This is not really a good option because it would be a format that can be only used for Sponge and would kinda defeat the points of a standardized mod metadata format. Would allow most control but has only downsides otherwise.)

**Note:** This is waiting on FML/Forge 1.8.4 for a decision currently.

**Comment by Azatik** [comment]
I like the third option.

**Comment by ST-DDT** [comment]
The third option. But we should use a different name such as `sponge.plugin` or `sponge.mod` or `spongemod.info`.

If something like that is added i will definitely use that.

**Comment by Mumfrey** [comment]
> This complicates things when using build systems because you either need to update the version manually in multiple places or write some hacky code that transforms the class and adds the properties to the annotation.

I disagree with this. Having the build system write the version into source using ant-style tokens or other special tokens (example [code](https://github.com/Vazkii/Botania/blob/master/src/main/java/vazkii/botania/common/lib/LibMisc.java#L20-L21) with [gradle lines](https://github.com/Vazkii/Botania/blob/master/build.gradle#L43-L44)) is pretty standard practice and incredibly easy to set up.

> * Don't support any metadata file: Build system integration would require transforming classes to inject the information in the annotation, which is obviously not a clean approach.

I don't really see how it's "not clean". As I said, replacing tokens in source is very straightforward and used in many build scripts.

> We should add a standardized way for Sponge plugins to provide the metadata optionally as a resource file. *These files are much easier to create in build systems*.

This is also untrue. Using a replace is pretty much equivalent to writing a file. In fact *it's exactly the same* because usually with a version file you're - guess what - reading a template file and replacing tokens in it!

This doesn't add anything, all you really do is introduce the possibility of having *conflicting* metadata when developers get confused and supply metadata in both locations.

**Comment by ST-DDT** [comment]
>> This complicates things when using build systems because you either need to update the version manually in multiple places or write some hacky code that transforms the class and adds the properties to the annotation.

> I disagree with this. Having the build system write the version into source using ant-style tokens or other special tokens (example code with gradle lines) is pretty standard practice and incredibly easy to set up.

Is there a short and simple Maven way to do it?
I found some huge pom snippets which copies the source to another folder, replaces the placeholders and then compiles it. But that does not look like a clean standarized way of doing that.
(If I would do that in production my (company's) certifier would kill me)

**Comment by Minecrell** [comment]
@Mumfrey 
> I disagree with this. Having the build system write the version into source using ant-style tokens or other special tokens (example code with gradle lines) is pretty standard practice and incredibly easy to set up.

The standard practice is (IMO) generally to use extra resource files for this. Replacing tokens in source code in this case is a specific feature added by ForgeGradle. For simple Sponge plugins, additional Gradle plugins shouldn't be necessary, and setting up token replacements in source files is (afaik) not supported natively in Gradle and Maven. To replace tokens in resource files, you can simply do it like this natively:

**Gradle:**
```gradle
import org.apache.tools.ant.filters.*

processResources {
    filter ReplaceTokens, tokens: [
        'project.version': project.version
    ]
}
```
**Maven:**
```pom
<resource>
    <directory>src/main/resources</directory>
    <filtering>true</filtering>
</resource>
```

There is nothing like `processSources` natively in Gradle and this is [similar for Maven](http://stackoverflow.com/q/2469922) from what I can tell. ForgeGradle [copies the whole sources to apply the replacements](https://github.com/MinecraftForge/ForgeGradle/blob/FG_1.2/src/main/java/net/minecraftforge/gradle/tasks/user/SourceCopyTask.java), and this also requires [quite a bit of code to hook into Gradle](https://github.com/MinecraftForge/ForgeGradle/blob/FG_1.2/src/main/java/net/minecraftforge/gradle/user/UserBasePlugin.java#L859-L909). Without a Gradle plugin this is a non-trivial amount of work for plugin authors to set up compared to resource replacements as shown above. With provided plugins for this feature, plugin authors still would need to figure out how to do this for other build systems.

**Comment by progwml6** [comment]
use mcmod.info as a large number of tools in the MC ecosystem support this format and use it to look for certain information. additionally, option 1 is not an option as it requires external tooling to have java code that scans annotations etc. to identify the plugin itself. Using option 2 requires only a handful of changes to tools that support gathering info for forge mods to work properly with sponge plugins instead of reinventing the wheel for NO reason whatsoever.

The only metadata that should be inside the code itself is the data needed to run the plugin properly. extra data such as author names, links should be inside metadata files.

everything else goes in mcmod.info

**Comment by progwml6** [comment]
there is no reason we can't add extra fields to mcmod.info that we find relevent and simply provide defaults or blank values for the few that are not relevent

**Comment by Minecrell** [comment]
@progwml6 I see your point with using `mcmod.info` but I disagree with the requirement to put additional metadata in a file instead of directly into the annotation. There is really no reason to require plugin authors to maintain this metadata in multiple places, and if they just want to make a simple plugin they should be able to set all their metadata in the plugin class directly.

Requiring this metadata file for additional information as authors or other things would just destroy the advantage of not having to create an additional resource file for plugins like in Bukkit for example.

**Comment by progwml6** [comment]
For people who are commenting on preferred options please give a reason for choosing option 2 or 3

**Comment by ST-DDT** [comment]
A more detailed explanation why i prefer option 3.

Basically Option 3 is only copy paste, with minor replacements.
Creating lists or similar things is quite easy.
In contrast to that using annotations for complex things like lists/arrays, becomes more and more messy the more lists you want to include, this also applies to required meta data.

For example i consider `required plugins`/`dependencies` a required meta data.

If you store it in a file everybody can read it. (as long as he can use a tool to unzip the jar)
In contrast to that all details saved in an annotation aren't readable at all. (unless you decompile it)
Some the main point is:
You store **all** plugin meta data in a **single**  file at a **fixed** location, that is **easily** accessible and **human** readable.

PS: I'm not sure but some plugins main classes might have a bad coding style so that you cannot even load them, because some dependencies might be missing. (java.lang.ExceptionInInitializerError) So you cannot even read the annotation, to display a useful error message why a specific plugin was unable to be loaded. Although the author added this to his `@Plugin` annotation.

**Comment by Minecrell** [comment]
@ST-DDT With your explanation, where is the difference between option 2 and 3 for you?

**Comment by progwml6** [comment]
@ST-DDT classes don't need to be loaded to check these ASM can be used to scan annotations

**Comment by ST-DDT** [comment]
@Minecrell 
The main difference between 2 and 3 is the dependency to an existing format.
If you want to be able to load a sponge plugin, you need the SpongeAPI and a reader that can read the `spongemod.info` file. So if teh API definition changes the implemenations have to adapt.

If we use mcmod.info, we have the dependency to forge, so if forge changes his file definition, the API has to accept the change (because its not the API's format specification). And all SpongeAPI implementations have to follow that although Sponge implementations cannot use the additional data provided by the new format at all.

Also i would like to add the `mainClass` to that file similar to Bukkit's `plugin.yml`, this helps me to debug buggy plugins and provide a fix or at least a detailed error description to the plugin author.

@progwml6 ASM doesn't look like the simpliest way to obtain that info reliably. Using `File`/`Resource` -> `Stream` -> `Reader` sounds easier for me. And if i'm capable of opening/extracting a jar i can use almost everything to read the plugin details/requirements.

**Comment by progwml6** [comment]
@ST-DDT to get the physical annotation info you will have a ton of trouble using those methods to get the info out of the bytecode.

we will NOT be adding a mainClass attribute it serves no point.

the mcmod.info format has had 1 minor change since i first heard about it around minecraft 1.3

**Comment by peterix** [comment]
Whatever you do guys:
make it authoritative and enforce format and correctness as much as possible. Otherwise, like the current `mcmod.info`, you'll end up with a **useless, misleading** file AND annotations.

**Comment by progwml6** [comment]
@peterix most of that is lazy developers. we are looking into populating some of the data automatically to keep user errors to a minimum in these files

**Comment by peterix** [comment]
I know. If it's done right from the start, it could be useful. But how it ended up with `mcmod.info` wasn't really useful for more than no guarantees display in a UI. It's not something I would feed to a package manager to figure out dependencies.

**Comment by ryantheleach** [comment]
I didn't like the plugin.yml that bukkit had, But if it helps Ore populate information and handle dependencies then maybe it's a better idea then doing it from code.

Otherwise there will need to be a whole toolchain rather then just let Ore handle it in any language it needs by reading from the archive.

**Comment by zml2008** [comment]
In maven I use a templating plugin to handle the issue of populating source files with build system info. See:
- [Usage in pom](https://github.com/zml2008/chatterbox/blob/master/pom.xml#L141)
- [Template file](https://github.com/zml2008/chatterbox/blob/master/src/main/java-templates/ninja/leaping/chatterbox/PomData.java)
- [Usage in annotation](https://github.com/zml2008/chatterbox/blob/master/src/main/java/ninja/leaping/chatterbox/ChatterBoxPlugin.java#L45)

This is the solution currently used by our [maven archetype](https://github.com/SpongePowered/sponge-plugin-archetype).

**Comment by MattBDev** [comment]
I would have to agree with @Mumfrey on this. Also, @zml2008 shows that this metadata isn't hard to populate using maven.

**Comment by boformer** [comment]
There should only be one way to set the metadata: A separate file or the annotation. 

It will confuse people if both variants are possible.

**Comment by Minecrell** [comment]
Ok, before we do any discussion on the format for the metadata file we should decide if we really want one. I think these requirements are clear:

- **We do _not_ want to remove the `@Plugin` annotation:** This has been in SpongeAPI since the first day and simplifies plugin creation quite a lot. Requiring plugin authors to add an additional metadata file would defeat the point of the annotation, so if we add it it needs to be either _optional_, or possible to be _generated automatically_ when building (e.g. using an annotation processor or Gradle/Maven plugin).

In my opinion we should add a file metadata format for the following reasons:

- **Unified (and simple) format to access plugin metadata for Minecraft mods/Sponge plugins:** This could be especially interesting together with MinecraftForge/FML#637. Plugin portals (such as [Ore](https://github.com/SpongePowered/Ore)) or custom launchers would benefit from an unified metadata file format working for most types of mods to display information about the author(s) or to download dependencies automatically for example.

  Reading the `@Plugin` annotation would only work for _Sponge plugins_, and would require bytecode parsing which can be quite complex (depending on the platform). A JSON file is easy to parse in almost all programming environments, and an unified format would even work for _different types of mods_, not only Sponge plugins.

- **Build system integration**: As mentioned above it is quite easy to setup source code token replacement, even though this is not supported natively by Gradle/Maven. The downside of this is that setting it up will always require manual setup of the tokens in the source code. Assuming someone wants to make a Gradle plugin which simplifies the creation of Sponge plugins, e.g.:

  ```gradle
  plugins {
      id 'com.example.sponge' version '1.0-SNAPSHOT'
  }

  // These should be automatically added to the plugin metadata
  name = 'ExamplePlugin'
  description = 'An example plugin'
  version = '1.0-SNAPSHOT'
  author = 'Minecrell'

  sponge {
      version = '2.1-SNAPSHOT'
  }
  ```

  With this, the user would still either need to setup all the tokens to be added in the source code manually, or the plugin would need to transform the bytecode generated by the compiler. Using an external file this would be only generating a simple JSON file.

**Comment by Minecrell** [comment]
Pinging @SpongePowered/developers for more opinions on this.

**Comment by progwml6** [comment]
for the record: i removed the example from the above post as to avoid confusion for developers until we pick an actual format

**Comment by modwizcode** [comment]
I think this is a good option, sometimes it's only metadata that needs a
fix and this saves a compile.

**Comment by progwml6** [comment]
@modwizcode some metadata will be in the @plugin annotation regardless

**Comment by killjoy1221** [comment]
If it's an issue of lazy developers, you can create the file for them using an annotation processor. These work regardless of build system (javac, maven/ant, gradle), but need to be enabled in the IDE. Luckily, you won't need to worry about that.  I'm sure you wouldn't have any trouble writing it seeing as you already have one for Mixins.

**Comment by MattBDev** [comment]
@Minecrell The issue you mention with build system is already partially solved with the maven archetype provided by sponge. @zml2008 mentioned this earlier. Not sure if you saw his comment or not...

**Comment by MattBDev** [comment]
> Reading the @Plugin annotation would only work for *Sponge plugins*
I don't see why it would be an issue that the @Plugin annotation in the *Sponge* API should be designed to work with only the implementations of the API. 
Trying to get developers to conform to a plugin/mod metadata file would be a pain in the ass considering there is already several formats already available and the fact that the @Plugin annotation has been the main way to specify a Plugin since the beginning....

**Comment by Mumfrey** [comment]
> If it's an issue of lazy developers, you can create the file for them using an annotation processor. These work regardless of build system (javac, maven/ant, gradle), but need to be enabled in the IDE. Luckily, you won't need to worry about that. I'm sure you wouldn't have any trouble writing it seeing as you already have one for Mixins.

For the record, I spoke to @AbrarSyed about this last week and agreed to make the AP, I just haven't had time.

**Comment by Minecrell** [comment]
@MattBDev 
> @Minecrell The issue you mention with build system is already partially solved with the maven archetype provided by sponge. `@zml2008` mentioned this earlier. Not sure if you saw his comment or not...

I mentioned this, however I've given an example which can not be (easily) solved using simple source code token replacement.

> I don't see why it would be an issue that the @Plugin annotation in the *Sponge* API should be designed to work with only the implementations of the API. 

The issue of `@Plugin` mentioned in my comment is not referring to implementations, the problem mentioned with reading the annotation is referring to general launchers or plugin portals like Ore which are not tied to Sponge plugins and would need to add special ways to get metadata for every separate metadata format. An unified format would simplify this a lot.

> Trying to get developers to conform to a plugin/mod metadata file would be a pain in the ass considering there is already several formats already available and the fact that the @Plugin annotation has been the main way to specify a Plugin since the beginning....

First part of the comment: we do **not** want to remove the `@Plugin` annotation and it won't be removed. We're basically looking for a way to allow developers to continue using _only_ the `@Plugin` annotation (that would generate the metadata file based on it using an annotation processor or similar) but still also allow for it to be provided manually.

For the format of the metadata file: one proposed way (see the issue description) is also to use the already quite well established `mcmod.info` format which we will definitely consider to use.

**Comment by MattBDev** [comment]
@Minecrell Thanks for the reply. I understand what you are looking for now. I think that an improved mcmod.info in JSON would be best. The best idea would be to work with Forge to aim for something similar to what is proposed in MinecraftForge/FML#637 so the same format (or similar with minor differences) can be used.

**Comment by TheRaspPie** [comment]
@Minecrell The problem with annotation processing is that it would only work if using the Java compiler. This means people who use Scala or any other JVM language would be required to provide a metadata file.

**Comment by varrix** [comment]
Any updates on this?

**Comment by Zidane** [comment]
@Minecrell 

I've read over the issue several times and here are my thoughts and what I want to happen:

* @Plugin annotation is staying. This is not up for discussion.
* Metafile is to become a requirement. Minecraft modding now goes for mcmod.info and we'll do the same. If the file isn't there we can generate it (based on what they've given in the @Plugin).
* @Minecrell its time to work your gradle magic to make it as painless as possible to generate this file. It has to be the bare minimum to put into a build.gradle for this (maybe a plugin?). Let me know what you have in mind. Ideally the file generated ends up in their final jar with the plugin dev not having to do really anything unless they desire to add more attributes.
* Will need something for maven folks but I don't do maven and we, as a project, as Gradle-based.

**Comment by ST-DDT** [comment]
I have some maven plugin experience.
If the requirements are clear I could implement this. Just ping me if I should start.

**Comment by TheRaspPie** [comment]
@Zidane This would clear all the possibility to use other JVM languages or build systems. It would be nice to have a small jar that generates the file if you run it. With this solution it would be easy to embed the jar into the build pipeline of any build system.

**Comment by Zidane** [comment]
@TheRaspPie 

The ease of making a plugin comes before "what other build systems will do". I await feedback on @Minecrell  to proceed.

**Comment by jk-5** [comment]
I have always been of the opinion that specifying some fields like version in the Plugin annotation is really annoying. It has been this way with forge and i share that opinion with the sponge api.
Having a Plugin/Mod annotation is a really nice system and should not go.

With the improved metadata format i suggest making specifying a version in the Plugin annotation optional or even not an option at all. Source code replacement is annoying and as far as i know not easy to do without a gradle plugin.

**Comment by EliterScripts** [comment]
Is this getting anywhere?

**Comment by Minecrell** [comment]
Fixed in https://github.com/SpongePowered/SpongeAPI/pull/1081
